async_fuse/src/memfs/dir.rs

```diff
@@ -1,46 +1,97 @@
 //! The implementation of directory related functionalities
 
-use nix::sys::stat::SFlag;
-use std::ffi::{CStr, OsStr, OsString};
+#[cfg(not(all(target_os = "linux", target_pointer_width = "64")))]
+compile_error!("async-fuse does not support this target now");
+
+use crate::util::{clear_errno, cstr_to_bytes, errno, with_c_str};
+
+use std::ffi::{OsStr, OsString};
+use std::io;
+use std::iter::FusedIterator;
 use std::os::unix::ffi::OsStrExt;
 use std::os::unix::io::RawFd;
-use std::ptr;
+use std::path::Path;
+use std::ptr::NonNull;
 
-use crate::fuse::protocol::INum;
+use libc::ino_t;
+use memchr::memchr;
+use nix::sys::stat::SFlag;
 
-#[cfg(target_os = "linux")]
-use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+/// Directory meta-data
+pub struct Dir(NonNull<libc::DIR>);
 
-#[cfg(target_os = "macos")]
-use libc::{dirent, readdir_r};
+unsafe impl Send for Dir {}
+unsafe impl Sync for Dir {}
 
-/// Directory meta-data
-#[derive(Debug)]
-pub struct Dir(ptr::NonNull<libc::DIR>);
+impl Drop for Dir {
+    fn drop(&mut self) {
+        let dirp = self.0.as_ptr();
+        unsafe {
+            let _ = libc::closedir(dirp);
```

Comment:
We'd better handle the error and log it if necessary.

Reply:
It an error occurs when dropping `Dir`, what is the expected behavior?
+ ignore?
+ panic?
+ debug_assert?
+ just log and continue?

Reply:
I recommend panic unless we know how to handle and recover.

Reply:
Panic inside `drop` will cause abort. The whole process will be terminated at once.
We have no chance to leave any information for debugging.